File: src/1169-Invalid-Transactions/1169.js
var invalidTransactions = function(transactions) {
    let data = {}, delt = {};
    for (let i in transactions) {
        let it = transactions[i].split(",");
        if (Number(it[2]) > 1000) {
            delt[i] = 0;
        }
        if (data[it[0]])
            data[it[0]].push([Number(it[1]), it[3], i]);
        else data[it[0]] = [[Number(it[1]), it[3], i]];
    }
    
    for (let k in data) {
        data[k].sort((a, b) => a[0] - b[0]);
        let q = [];
        for (let it of data[k]) {
            while (q.length > 0 && q[0][0] < it[0] - 60) {
                q.shift();
            }
            for (let ix of q) {
                if (it[1] != ix[1]) {
                    delt[it[2]] = 0;
                    delt[ix[2]] = 0;
                }
            }
            q.push(it);
        }
    }
    let res = [];
    for (let k in delt) {
        res.push(transactions[k]);
    }
    return res;
};

File: src/1169-Invalid-Transactions/1169.go
type tx struct {
    time int
    city string
    idx int
}

func invalidTransactions(transactions []string) []string {
    data := make(map[string][]tx)
    delt := make(map[int]int)
    for i, v := range transactions {
        it := strings.Split(v, ",")
        t, _ := strconv.Atoi(it[1])
        amout, _ := strconv.Atoi(it[2])
        if amout > 1000 {
            delt[i] = 0
        }
        data[it[0]] = append(data[it[0]], tx{time: t, city: it[3], idx: i})
    }
    
    for v, _ := range data {
        sort.Slice(data[v], func(i, j int) bool {
            return data[v][i].time < data[v][j].time
        })
        q := make([]tx, 0)
        for _, it := range data[v] {
            for len(q) > 0 && q[0].time < it.time - 60 {
                q = q[1:]
            }
            for _, ix := range q {
                if it.city != ix.city {
                    delt[it.idx], delt[ix.idx] = 0, 0
                }
            }
            q = append(q, it)
        }
    }
    res := []string{}
    for i, _ := range delt {
        res = append(res, transactions[i])
    }
    return res
}